src/utils/bindtwostepcarettoattribute.js

```diff
@@ -0,0 +1,139 @@
+ * For licensing, see LICENSE.md.
+ */
+
+/**
+ * @module engine/utils/bindtwostepcarettoattribute
+ */
+
+import { keyCodes } from '@ckeditor/ckeditor5-utils/src/keyboard';
+
+/**
+ * This helper adds two-steps caret movement behaviour for given attribute.
+ *
+ * When caret is moving by arrow keys and reach bound of text with attribute for which behaviour is enabled
+ * then typing does not expand this attribute. Additional arrow key press is needed to "enter" to the text
+ * and start typing with this attribute. The same is is for leaving this text.
+ *
+ * When behaviour is enabled for `linkHref` attribute and caret is just before the attribute element then pressing
+ * right arrow will move caret inside the attribute element instead of moving after next character:
+ *
+ * 		<p>foo{}<a>bar</a>biz<p> `->` <p>foo<a>{}foo</a>barr<p>
+ *
+ * The same is for "leaving" attribute element:
+ *
+ * 		<p>foo<a>bar{}</a>biz<p> `->` <p>foo<a>bar</a>{}biz<p>
+ *
+ * And when moving left:
+ *
+ * 		<p>foo<a>bar</a>{}biz<p> `<-` <p>foo<a>bar{}</a>biz<p>
+ * 		<p>foo<a>{}bar</a>biz<p> `<-` <p>foo{}<a>bar</a>biz<p>
+ *
+ * @param {module:engine/view/view~View} view View controller instance.
+ * @param {module:engine/model/model~Model} model Data model instance.
+ * @param {module:utils/dom/emittermixin~Emitter} emitter The emitter to which this behavior should be added.
+ * @param {String} attribute Attribute for which behaviour will be added.
+ */
+export default function bindTwoStepCaretToAttribute( view, model, emitter, attribute ) {
+	const modelSelection = model.document.selection;
+
+	// Listen to keyboard events and handle cursor before the move.
+	emitter.listenTo( view.document, 'keydown', ( evt, data ) => {
+		const arrowRightPressed = data.keyCode == keyCodes.arrowright;
+		const arrowLeftPressed = data.keyCode == keyCodes.arrowleft;
+
+		// When neither left or right arrow has been pressed then do noting.
+		if ( !arrowRightPressed && !arrowLeftPressed ) {
+			return;
+		}
+
+		// This implementation works only for collapsed selection.
+		if ( !modelSelection.isCollapsed ) {
+			return;
+		}
+
+		// User tries to expand selection, so two-steps movement is not necessary.
+		if ( data.shiftKey ) {
+			return;
+		}
+
+		const position = modelSelection.getFirstPosition();
+
+		// Moving right.
+		if ( arrowRightPressed ) {
+			// If gravity is already overridden then do nothing.
+			// It means that we already enter `foo<a>{}bar</a>biz` or left `foo<a>bar</a>{}biz` text with attribute
+			// and gravity will be restored just after caret movement.
+			if ( modelSelection.isGravityOverridden ) {
+				return;
+			}
+
+			// If caret sticks to the bound of Text with attribute it means that we are going to
+			// enter `foo{}<a>bar</a>biz` or leave `foo<a>bar{}</a>biz` the text with attribute.
+			if ( isStickToAttribute( position.nodeAfter, position.nodeBefore, attribute ) ) {
```

Comment:
stick is a verb so `sticksToAttribute()` or something like `atAttributeBoundary()`?